horovod/common/mpi_context.cc

```diff
@@ -77,5 +85,120 @@ int MPIContext::GetMPITypeSize(DataType dtype) {
   return out;
 }
 
+void MPIContext::Initialize(const std::vector<int>& ranks,
+                            MPIContextManager* ctx_manager) {
+  // Initialize MPI if it was not initialized. This must happen on the
+  // background thread, since not all MPI implementations support being called
+  // from multiple threads.
+  //
+  // In some cases MPI library has multi-threading support, but it slows down
+  // certain components, e.g. OpenIB BTL in OpenMPI gets disabled if
+  // MPI_THREAD_MULTIPLE is requested.
+  //
+  // By default, we will ask for multiple threads, so other libraries like
+  // mpi4py can be used together with Horovod if multi-threaded MPI is
+  // installed.
+  auto mpi_threads_disable = std::getenv(HOROVOD_MPI_THREADS_DISABLE);
+  int required = MPI_THREAD_MULTIPLE;
+  if (mpi_threads_disable != nullptr &&
+      std::strtol(mpi_threads_disable, nullptr, 10) > 0) {
+    required = MPI_THREAD_SINGLE;
+  }
+  int provided;
+  int is_mpi_initialized = 0;
+  MPI_Initialized(&is_mpi_initialized);
+  if (is_mpi_initialized) {
+    MPI_Query_thread(&provided);
```

Comment:
Let's move `provided` into this scope since it's only used here.

Reply:
Yes.